Add 'noarch' tests which only run in one CI config

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #53973 Add some trivial meta functions for commonly used inplace operations
* #53972 Make storage access error NotImplementedError
* #53823 Compute type_equal() without reference to backend()
* #53816 Also propagate storage_access_should_throw_ when copying tensor metadata
* #53759 Fix restriding logic for structured kernels
* #53682 Add meta device to generic device testing framework, skip NotImplementedError
* **#53747 Add 'noarch' tests which only run in one CI config**

Fixes #53743

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D26971343](https://our.internmc.facebook.com/intern/diff/D26971343)

ezyang committed:
Add 'noarch' tests which only run in one CI config

Fixes #53743

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

mruberry commented on torch/testing/_internal/common_utils.py:
```diff
@@ -572,6 +577,16 @@ def wrapper(*args, **kwargs):
     return wrapper
 
 
+def noarchTest(fn):
```
Comment for this decorator explaining what a "NoArch" test is and how the decorator is intended to be used? 

What's the goal of running these tests on only one architecture? If we want to run all meta tests on only a single architecture, for example, then we could just decline to add the Meta device to the list of available devices on CI builds other than the designated one.

mruberry reviewed (approved):
This looks correct up to name bikeshedding (which we should probably use the issue for)

mruberry commented:
Test failures appear to be in the base.

ezyang committed:
Update on "Add 'noarch' tests which only run in one CI config"


Fixes #53743

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

janeyx99 commented:
This seems like a functional way to get the intended behavior, but I have some uneasiness regarding just picking a config and having that somewhat "silently" be designated the noarch config.

Concerns:
- Despite the WARNING and the comments throughout the code regarding noarch tests, it is still possible for someone to mess up and delete the config without moving the noarch tests, and we have no fast obvious way of detecting if that happens.
- If for whatever reason this build/test config was corrupted and couldn't run for a commit, it isn't immediately clear that the noarch tests may not have been run at all.

Potential solution that may alleviate the above: 
I think I'd be more comfortable if we had a designated test config that only ran noarch tests (similar to how we have a separate slow tests CI job). The test config can stem from the `pytorch-linux-bionic-py3.6-clang9` build. If we have "noarch" in the test CI job name/build environment, it'd be more clear when the job gets removed from CI later on. It'd also isolate the noarch tests from other configs so we'd get less mixed signal.

janeyx99 commented on .jenkins/pytorch/test.sh:
```diff
@@ -26,6 +26,12 @@ if [[ "$BUILD_ENVIRONMENT" == *cuda11* ]]; then
   export BUILD_SPLIT_CUDA=ON
 fi
 
+if [[ "$BUILD_ENVIRONMENT" == *pytorch-linux-bionic-py3.6-clang9* ]]; then
```
An advantage to using the potential solution I mentioned in the other comment is we'd be able to have something more clear here, like: 
```suggestion
if [[ "$BUILD_ENVIRONMENT" == *noarch* ]]; then
```

janeyx99 commented on test/test_torch.py:
```diff
@@ -2581,13 +2581,15 @@ def test_batch_norm_cpu_inference(self):
                     self.assertEqual(output3, output1)
                     self.assertEqual(output3, output2)
 
+        @noarchTest
```
Will there also be a PR going through every test in the codebase so far to mark some as noarch or will it be more of a gradual "from now on" thing?

ezyang replied:
It'll be as needed. The main motivating use case for this is I'm adding 17k new tests in #53682 and I don't want them to run on all CI runs.

janeyx99 commented on torch/testing/_internal/common_utils.py:
```diff
@@ -572,6 +577,20 @@ def wrapper(*args, **kwargs):
+    @wraps(fn)
+    def wrapper(*args, **kwargs):
+        if TEST_SKIP_NOARCH:
+            raise unittest.SkipTest("test is noarch but we are skipping noarch tests due to TEST_SKIP_NOARCH")
```
nitty nit (the "but" was a little confusing but it's ok the way it is now too if you want to land)
```suggestion
            raise unittest.SkipTest("test is noarch: we are skipping noarch tests due to TEST_SKIP_NOARCH")
```

ezyang committed:
Update on "Add 'noarch' tests which only run in one CI config"


Fixes #53743

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D26971343](https://our.internmc.facebook.com/intern/diff/D26971343)

[ghstack-poisoned]

ezyang committed:
Update on "Add 'noarch' tests which only run in one CI config"


Fixes #53743

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D26971343](https://our.internmc.facebook.com/intern/diff/D26971343)

[ghstack-poisoned]

ezyang committed:
Update on "Add 'noarch' tests which only run in one CI config"


Fixes #53743

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D26971343](https://our.internmc.facebook.com/intern/diff/D26971343)

[ghstack-poisoned]

ezyang committed:
Update on "Add 'noarch' tests which only run in one CI config"


Fixes #53743

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D26971343](https://our.internmc.facebook.com/intern/diff/D26971343)

[ghstack-poisoned]

ezyang commented:
@janeyx99 okey dokey, the CI job has been renamed (I thought this would be painful but actually we have ignored for docker tags so it's fine)

janeyx99 commented on .circleci/cimodel/data/pytorch_build_definitions.py:
```diff
@@ -316,6 +317,9 @@ def instantiate_configs():
             python_version = fc.find_prop("pyver")
 
+        if is_noarch:
+            parms_list_ignored_for_docker_image.append("noarch")
```
```suggestion
           restrict_phases = ["build"] if restrict_phases is None else restrict_phases
           restrict_phases.extend(["test", "noarch_test"])
```
I believe this code would not change the name of the existing build but instead would create a separate CI test job that only runs noarch tests, which would help with isolating the tests, if that is to be desired

ezyang replied:
I think we don't want this change? I don't want to split out the noarch tests into one config that only runs them, because I don't have any evidence that sharding them separately would be a benefit (noarch is much faster than normal tests so it would be super lopsided).

janeyx99 replied:
Yea, that's fine in that case.

ezyang committed:
Update on "Add 'noarch' tests which only run in one CI config"


Fixes #53743

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D26971343](https://our.internmc.facebook.com/intern/diff/D26971343)

[ghstack-poisoned]

ezyang committed:
Update on "Add 'noarch' tests which only run in one CI config"


Fixes #53743

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D26971343](https://our.internmc.facebook.com/intern/diff/D26971343)

[ghstack-poisoned]

ezyang committed:
Update on "Add 'noarch' tests which only run in one CI config"


Fixes #53743

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D26971343](https://our.internmc.facebook.com/intern/diff/D26971343)

[ghstack-poisoned]
